onadata/libs/utils/csv_import.py

```diff
@@ -470,3 +371,229 @@ def submission_xls_to_csv(xls_file):
         csv_writer.writerow(row_values)
 
     return csv_file
+
+
+def validate_csv(csv_file, xform):
+    """Validate a CSV data being imported to an existing form
+
+    Takes a csv formatted file or sting containing rows of submission/instance
+    and validates the data making sure that the data is valid enough to be
+    submitted depending on the unique contraints of the existing form.
+
+    :param onadata.apps.logger.models.XForm xform: The subbmission's XForm.
+    :param (str or file): A CSV formatted file with submission rows.
+    :return: Returns a dict with the validation status
+    and data is returned.
+    :rtype: Dict
+    """
+    # Validate csv_file is utf-8 encoded or unicode
+    if isinstance(csv_file, str):
+        csv_file = BytesIO(csv_file)
+    elif csv_file is None or not hasattr(csv_file, 'read'):
+        raise Exception(
+            u'Invalid param type for csv_file`.'
+            'Expected utf-8 encoded file or unicode'
+            ' string got {} instead'.format(type(csv_file).__name__))
+
+    # Change stream position to start of file
+    csv_file.seek(0)
+
+    csv_reader = ucsv.DictReader(csv_file, encoding='utf-8-sig')
+    csv_headers = csv_reader.fieldnames
+
+    # Make sure CSV headers have no spaces
+    if any(' ' in header for header in csv_headers):
```

Comment:
Excel doesn't support spaces within its headers afaik. And adding an underscore to the header will not necessarily help much in this case if the `Ona` form doesn't have a matching header in the `XForm` the csv is being imported too.

Reply:
From this thread [here](https://answers.microsoft.com/en-us/office/forum/office_2010-excel/how-do-i-add-special-characters-in-excels-header/ce70abab-2df7-4125-bbdc-26c64512220e), I'm assuming special characters are not a problem in excel. Maybe @ukanga can share some more light into this.

Reply:
> interesting, do you have a reference for spaces not being supported?

https://community.talend.com/t5/Design-and-Development/resolved-Spaces-in-the-column-headers-of-Excel-output-file/td-p/104454

https://support.office.com/en-us/article/using-structured-references-with-excel-tables-f5ed2452-2337-4f71-bed3-c8ae6d2b276e --> Under `How do i change table name`

Reply:
Hmm, this isn't clear to me, this document is referring to "table names" which sound different from column names. They have what look like columns in it that contain spaces, like "Sales Amount".

Did you try not replacing spaces and see that fail? Is there a test in our code for column names with spaces? You could see if that fails without conversation, if you haven't checked this already

Reply:
Sorry about that, was operating on the wrong reference. Their is a test for the csv spaces that fails when csv headers have spaces in it, can be found [here](https://github.com/onaio/onadata/blob/b3d35ec710dc0abc5f6a150f6724fc19d6ddc13a/onadata/libs/tests/utils/test_csv_import.py#L161). It was implemented due to this issue [here](https://github.com/onaio/onadata/issues/693)

Reply:
My bad seems like i was following the wrong reference. There is a test for spaces in headers [here](https://github.com/onaio/onadata/blob/f378fde24bbe93215ce56efee2d36cc93f28eac1/onadata/libs/tests/utils/test_csv_import.py#L161), it was implemented to solve this issue [here](https://github.com/onaio/onadata/issues/695). Due to the [XML naming rules](https://www.w3schools.com/xml/xml_elements.asp).

onadata also doesn't allow spaces in headers while uploading a form:
<img width="658" alt="Screenshot 2019-12-13 at 09 34 17" src="https://user-images.githubusercontent.com/25849009/70777861-e359d480-1d91-11ea-8eab-5bd4e2ce7134.png">

It fails validation process due to missing header when if I keep an underscore to replace the space.

Reply:
Replacing the space with a `_` works in the case above ^ but the field could be something like `DATE.OUVERTURE` or `DATE-OUVERTURE` too.

Reply:
OK, thanks for the details that makes sense. Can you add a comment to the code to make this clear for future readers, like adding onto existing comment `... because these are converted to XLSForm names which cannot have spaces.`

Reply:
Added the comment.